lib/hal/audio.c

```diff
@@ -109,26 +148,48 @@ void XAudioInit(int sampleSizeInBits, int numChannels, XAudioCallback callback,
 	MmLockUnlockBufferPages((PVOID)pac97device, sizeof(AC97_DEVICE), FALSE);
 
 	pac97device->mmio = (unsigned int *)0xfec00000;
-	pac97device->nextDescriptorMod31 = 0;
+	pac97device->nextDescriptor = 0;
 	pac97device->callback = callback;
 	pac97device->callbackData = data;
 	pac97device->sampleSizeInBits = sampleSizeInBits;
 	pac97device->numChannels = numChannels;
 
-	// initialise descriptors to all 0x00 (no samples)        
-	memset((void *)&pac97device->pcmSpdifDescriptor[0], 0, sizeof(pac97device->pcmSpdifDescriptor));
-	memset((void *)&pac97device->pcmOutDescriptor[0], 0, sizeof(pac97device->pcmOutDescriptor));
+	volatile unsigned char *pb = (unsigned char *)pac97device->mmio;
+
+	// initialise descriptors to all zero (no samples)
+	for(unsigned int i = 0; i < 32; i++) {
+		pac97device->pcmSpdifDescriptor[i].bufferStartAddress = 0;
+		pac97device->pcmSpdifDescriptor[i].bufferLengthInSamples = 0;
+		pac97device->pcmSpdifDescriptor[i].bufferControl = 0;
+		pac97device->pcmOutDescriptor[i].bufferStartAddress = 0;
+		pac97device->pcmOutDescriptor[i].bufferLengthInSamples = 0;
+		pac97device->pcmOutDescriptor[i].bufferControl = 0;
+	}
 
-	// perform a cold reset
+	// perform cold reset
+	pac97device->mmio[0x12C>>2] &= ~2;
+	LARGE_INTEGER Interval;
+	Interval.QuadPart = -10;
+	KeDelayExecutionThread(KernelMode, FALSE, &Interval);
 	pac97device->mmio[0x12C>>2] |= 2;
 	
 	// wait until the chip is finished resetting...
 	while(!(pac97device->mmio[0x130>>2]&0x100))
 		;
 
+	// reset bus master registers for analog output
+	pb[0x11B] = (1 << 4) | (1 << 3) | (1 << 2) | (1 << 1);
+	while(pb[0x11B] & (1 << 1))
+		;
+
+	// reset bus master registers for digital output
+	pb[0x17B] = (1 << 4) | (1 << 3) | (1 << 2) | (1 << 1);
```

Comment:
This can be done  at the same time as the other reset, to reduce wait times

Reply:
I'm not planning to change this, because I don't want to force another round of testing.
I'll need to look at AC97 for XQEMU anyway, and I can clean this up then.
During a cleanup pass, I might also define constants for these fields etc. (my code just mirrors the existing style - which was also ugly), if I don't end up making an all new API altogether.